components/script/dom/servoparser/async_html.rs

```diff
@@ -20,218 +21,249 @@ use dom::htmltemplateelement::HTMLTemplateElement;
+        let (to_sink_sender, sink_receiver) = channel();
+        // Messages from the Sink (parser thread) to Tokenizer (main thread)
+        let (sink_sender, from_sink_receiver) = channel();
+
+        let mut tokenizer = Tokenizer {
             document: JS::from_ref(document),
-            current_line: 1,
-            script: Default::default(),
-            parse_node_data: HashMap::new(),
-            next_parse_node_id: Cell::new(1),
+            html_tokenizer_receiver: from_html_tokenizer_receiver,
+            html_tokenizer_sender: to_html_tokenizer_sender,
+            sink_receiver: from_sink_receiver,
+            sink_sender: to_sink_sender,
             nodes: HashMap::new(),
-            document_node: ParseNode {
-                id: 0,
-                qual_name: None,
-            }
+            url: url
         };
-        let data = ParseNodeData::default();
-        sink.insert_parse_node_data(0, data);
-        sink.insert_node(0, JS::from_ref(document.upcast()));
-        sink
+        tokenizer.insert_node(0, JS::from_ref(document.upcast()));
+
+        let mut sink = Sink::new(sink_sender, sink_receiver);
+        let mut ctxt_parse_node = None;
+        let mut form_parse_node = None;
+        let mut fragment_context_is_some = false;
+        if let Some(fc) = fragment_context {
+            let node = sink.new_parse_node();
+            tokenizer.insert_node(node.id, JS::from_ref(fc.context_elem));
+            ctxt_parse_node = Some(node);
+
+            form_parse_node = fc.form_elem.map(|form_elem| {
+                let node = sink.new_parse_node();
+                tokenizer.insert_node(node.id, JS::from_ref(form_elem));
+                node
+            });
+            fragment_context_is_some = true;
+        };
+
+        // Create new thread for HtmlTokenizer. This is where parser actions
+        // will be generated from the input provided. These parser actions are then passed
+        // onto the main thread to be executed.
+        thread::Builder::new().name(String::from("HTML Parser")).spawn(move || {
+            run(sink,
+                fragment_context_is_some,
+                ctxt_parse_node,
+                form_parse_node,
+                html_tokenizer_sender,
+                html_tokenizer_receiver);
+        }).expect("HTML Parser thread spawning failed");
+
+        tokenizer
     }
 
-    fn new_parse_node(&mut self) -> ParseNode {
-        let id = self.next_parse_node_id.get();
-        let data = ParseNodeData::default();
-        self.insert_parse_node_data(id, data);
-        self.next_parse_node_id.set(id + 1);
-        ParseNode {
-            id: id,
-            qual_name: None,
+    #[allow(unsafe_code)]
+    pub fn feed(&mut self, input: &mut BufferQueue) -> Result<(), Root<HTMLScriptElement>> {
+        enum Message {
+            Sink(FromSinkMsg),
+            HtmlTokenizer(FromHtmlTokenizerMsg)
+        }
+
+        let mut send_tendrils = Vec::new();
```

Comment:
Instead of making the BufferQueue implement Send, I now extract the `StrTendril`s from the BufferQueue and convert them to `SendTendril`s, which are then stored in a mutex. After we get a result from the feed function, we update `input` based on the contents of the mutex. The mutex is updated after `HtmlTokenizer`'s feed method returns.